lib/web/users.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/web/ui"
+	"github.com/gravitational/trace"
+
+	"github.com/julienschmidt/httprouter"
+)
+
+// requestUser is used to unmarshal JSON requests.
+// Requests are made from the web UI for:
+//	- user creation
+type requestUser struct {
+	Name  string   `json:"name"`
+	Roles []string `json:"roles"`
+}
+
+// responseCreateUser is used to send back data
+// about created local user and password setup token.
+type responseCreateUser struct {
+	User  ui.User       `json:"user"`
+	Token ui.ResetToken `json:"token"`
+}
+
+// responseGetUsers is used to send back list of all locally saved users.
+type responseGetUsers struct {
+	Users []ui.User `json:"users"`
+}
+
+// createUser allows a UI user to create a new user.
+//
+// POST /webapi/sites/:site/namespaces/:namespace/users
+//
+// Request:
+// {
+//		"username": "foo",
+//		"roles": ["role1", "role2"]
+// }
+//
+// Response:
+// {
+//		"user": {...},
+//		"token": {...}
+// }
+func (h *Handler) createUser(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	// Pull out request data.
+	var req *requestUser
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := req.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Check user doesn't exist already.
```

Comment:
Could you move this check to `CreateUser` method? The reason why we had to do it this way was because of "UpsertUser" API  which we replaced with CreateUser and UpdateUser.

webapi needs to call 

```golang
user, err := ctx.clt.CreateUser(r.Context(), user)
```

and does not need to check if the user exists.

Reply:
moved this check to `CreateUser` method